howso/residuals.amlg

```diff
@@ -129,62 +181,127 @@
 	(set
 		feature_residuals_map
 		"residual_map"
-		(map
-			(lambda
-				;set upper bound to nominals to be max nominal deviation, accounting for imbalanced classes
-				(if (contains_index !nominalsMap (current_index))
-					(let
-						(assoc
-							;map of class -> count
-							class_counts_map
-								(compute_on_contained_entities (list
-									(query_value_masses (current_index 2) (null) (contains_index !numericNominalFeaturesMap (current_index 2)) )
-								))
-						)
-
-						;do not max cap nominal deviations if less than 2 classes have been trained
-						(if (<= (size class_counts_map) 1)
-							(max
-								(get !cachedFeatureMinResidualMap (current_index))
-								(current_value)
-								(get !userSpecifiedFeatureErrorsMap (current_index))
-							)
-
-							;else cap the max nominal deviation
+		(let (assoc
+			temp_residual_map
+				(map
+					(lambda
+						;set upper bound to nominals to be max nominal deviation, accounting for imbalanced classes
+						(if (contains_index !nominalsMap (current_index))
 							(let
-								(assoc total_count (apply "+" (values class_counts_map)) )
+								(assoc
+									;map of class -> count
+									feature_grouping
+										(if (size !sharedDeviationsMap)
+											(call !GetSharedDeviationGrouping (assoc feature_group_to_retrieve (current_index 2)))
+											(null)
+										)
+									current_feature (current_index 1)
+								)
+								(declare (assoc
+									class_counts_map
+										(if (size feature_grouping)
+											;get the class counts for all the features in the shared deviations group
+											(reduce
```

Comment:
this block of code does not work at all
a) this always results in a null, because if you use a reduce, the code needs to be in a (lambda) like other iterators
b) doing an (apply "append" will append the assocs, but if they have same indices/classes they will be clobbered, eg {a:5, b:4} appended with {a:3, c::2} results in {a:3, b:4, c::2} and not the expected {a:8, b:4, c:2}.  we'll need to rethink this